### PR TITLE
refactor(primary-action-dropdown): use hooks

### PR DESCRIPTION
**Pull request by montezume**
#### Summary

Refactors the component `PrimaryActionDropdown` to use hooks

### CODE REVIEW

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js [hunk 1]
```diff
@@ -211,89 +208,82 @@ Option.defaultProps = {
...
+      }
+    },
+    [ref]
+  );
 
-  handleClose = () => {
-    this.setState({ isOpen: false });
-  };
+  React.useEffect(() => {
+    window.addEventListener('click', handleGlobalClick);
```
Comment (montezume) [inline src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: this replaces the `componentDidMount` and `componentWillUnmount`

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.spec.js [hunk 2]
```diff
@@ -142,8 +143,10 @@ describe('when dropdown is open and body is clicked', () => {
     expect(getByLabelText('Secondary Action')).toBeInTheDocument();
 
     // click on doucment body
-    document.body.click();
 
+    act(() => {
```
Comment (montezume) [inline src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.spec.js]: needed this or else got warnings in the test

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js [hunk 3]
```diff
@@ -211,89 +208,82 @@ Option.defaultProps = {
...
+    return () => {
+      window.removeEventListener('click', handleGlobalClick);
+    };
+  }, [handleGlobalClick]);
```
Comment (tdeekens) [inline src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: I dont even think this has to be a dep. Dep talk all hook lomg.

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js [hunk 4]
```diff
@@ -211,89 +208,82 @@ Option.defaultProps = {
...
-
-  static propTypes = {
-    children: PropTypes.node.isRequired,
-  };
-
-  state = {
-    isOpen: false,
-  };
+const PrimaryActionDropdown = props => {
+  const ref = React.useRef();
```
Comment (tdeekens) [inline src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: Can we name the hook better?

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.spec.js (hunk 2 above)
Comment (tdeekens) [reply src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.spec.js]: Cool that it worked.

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js [hunk 5]
```diff
@@ -230,7 +230,7 @@ const PrimaryActionDropdown = props => {
     return () => {
       window.removeEventListener('click', handleGlobalClick);
     };
-  }, [handleGlobalClick]);
+  });
```
Comment (tdeekens) [inline src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: We want it to run once not on each render right. So empty array?

Comment (montezume) [reply src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: good call

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js [hunk 6]
```diff
@@ -211,89 +208,82 @@ Option.defaultProps = {
...
+    event => {
+      const dropdownButton = ref.current;
+      if (
+        event.target !== dropdownButton &&
+        !dropdownButton.contains(event.target)
+      ) {
+        setIsOpen(false);
+      }
+    },
+    [ref]
```
Comment (emmenko) [inline src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: I think I saw an issue or something related to defining DOM elements (refs) as dependencies. I’ll see if I can find more about it but maybe we should double check if it makes sense to do so.

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js [hunk 7]
```diff
@@ -211,89 +208,82 @@ Option.defaultProps = {
...
-
-  handleOpen = () => {
-    this.setState({ isOpen: true });
-  };
+  const handleGlobalClick = React.useCallback(
+    event => {
+      const dropdownButton = ref.current;
+      if (
+        event.target !== dropdownButton &&
+        !dropdownButton.contains(event.target)
```
Comment (emmenko) [inline src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: Should we also check if `dropdownButton` is defined? 🤔

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js [hunk 8]
```diff
@@ -211,89 +208,82 @@ Option.defaultProps = {
...
-  static propTypes = {
-    children: PropTypes.node.isRequired,
-  };
-
-  state = {
-    isOpen: false,
-  };
+const PrimaryActionDropdown = props => {
+  const ref = React.useRef();
+  const [isOpen, setIsOpen] = React.useState(false);
```
Comment (emmenko) [inline src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: Use the new collapsible/toggle hook?

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js [hunk 9]
```diff
@@ -211,89 +208,82 @@ Option.defaultProps = {
...
+      css={css`
+        position: relative;
+        display: inline-flex;
+        align-items: column;
+      `}
+    >
+      <DropdownHead
+        iconLeft={primaryOption.props.iconLeft}
+        isDisabled={primaryOption.props.isDisabled}
+        onClick={isOpen ? () => setIsOpen(false) : primaryOption.props.onClick}
```
Comment (emmenko) [inline src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: Nitpick: memoized callback?

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js [hunk 10]
```diff
@@ -211,89 +208,82 @@ Option.defaultProps = {
...
+        chevron={
+          <DropdownChevron
+            ref={ref}
+            onClick={
```
Comment (emmenko) [inline src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: Same here?

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.spec.js (hunk 2 above)
Comment (emmenko) [reply src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.spec.js]: Since RTL already includes this, can you maybe try to click on the rendered container?

```js
fireEvent.click(rendered.container)
```

There should be no need for using `act` directly

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js (hunk 8 above)
Comment (montezume) [reply src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: sure, but need to merge that first 😉

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.spec.js (hunk 2 above)
Comment (emmenko) [reply src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.spec.js]: Fixed here: https://github.com/commercetools/ui-kit/pull/965/commits/b83d74f503cd5d44322219ca1b0b6b2a7c2f624f

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js (hunk 8 above)
Comment (emmenko) [reply src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: Actually we can't, because in the `handleGlobalClick` we force to be `false`, so it's not really a toggle anymore

File: src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js [hunk 11]
```diff
@@ -211,89 +208,89 @@ Option.defaultProps = {
...
-  };
+  const childrenAsArray = React.Children.toArray(props.children);
+  const primaryOption =
+    childrenAsArray.find(option => !option.props.isDisabled) ||
+    childrenAsArray[0];
 
-  componentDidMount() {
-    window.addEventListener('click', this.handleGlobalClick);
-  }
+  const { onClick } = primaryOption.props;
```
Comment (emmenko) [inline src/components/dropdowns/primary-action-dropdown/primary-action-dropdown.js]: Same drill here again

File: src/index.js [hunk 12]
```diff
@@ -141,3 +140,5 @@ export { default as version } from './version';
 export {
   default as ContentNotification,
 } from './components/notifications/content-notification';
+
+export * from './hooks';
```
Comment (emmenko) [inline src/index.js]: Extra unrelated little change.

Comment (tdeekens) [reply src/index.js]: Does UIKit really want to export it's hooks already? This would maybe even a feature releasea as a result.

Comment (emmenko) [reply src/index.js]: It was already exposed actually, I just refactored it to `export *` instead of `export { default as useToggleState }`

Comment (tdeekens) [reply src/index.js]: Hm. I, personal opinion ahead, liked it before more then. The library opts into what it wants to expose instead of defaulting to everything.

Comment (emmenko) [reply src/index.js]: Sure my bad, I'll revert this change then

Comment (emmenko) [reply src/index.js]: https://github.com/commercetools/ui-kit/pull/965/commits/8361bc17e7c9cd7d84db70adf633bceb0d1c5821
